lib/sorbet-coerce.rb

```diff
@@ -1,22 +1,31 @@
 # typed: strict
+require_relative 'configuration'
 require 'private/converter'
 require 'safe_type'
 
-module T
-  class CoercionError < SafeType::CoercionError; end
+module T::Coerce
+  class ValueError < SafeType::CoercionError; end
+  class ShapeError < SafeType::CoercionError; end
 
-  module Coerce
-    define_singleton_method(:[]) do |type|
-      Class.new(T::Private::Converter) do
-        define_method(:to_s) { "#{name}#[#{type.to_s}]" }
+  define_singleton_method(:[]) do |type|
+    Class.new(T::Private::Converter) do
+      define_method(:to_s) { "#{name}#[#{type.to_s}]" }
 
-        define_method(:from) do |args|
-          begin
-            T.send('let', send('_convert', args, type), type)
-          rescue TypeError
-            raise CoercionError.new(args, type)
-          end
+      define_method(:from) do |args, raise_value_error: nil|
+        if raise_value_error.nil?
+          raise_value_error = T::Coerce::Configuration.raise_value_error
         end
+
+        error = nil
+        begin
+          return T.send('let', send('_convert', args, type, raise_value_error), type)
+        rescue TypeError
```

Comment:
This part feels weird to me. We probably shouldn't turn TypeError into ValueError. Do we even need to call `T.let` here?

Reply:
> We probably shouldn't turn TypeError into ValueError. Do we even need to call T.let here?


> I think we should change "ValueError" to CoerceError for clarity. If it is TypeError people know it comes from Sorbet. We only raise the errors we know of.

Ah, that makes a lot of sense. I like this handling better!